### PR TITLE
BLD: update merge script to update on github

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
@jreback @TomAugspurger This adds the ability to checkout a PR, do a small fixup or rebase, and push changes to github. 
I now added it as an additional question whether to update or merge, but can also make to separate scripts of it for convenience.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Just to make sure, the goal of this is to allow you to push your local changes back up to the PR branch, correct? And then use the green "Squash and merge" from there?

IIUC the new options are

1. Update the PR (`y`)
2. Merge the PR (maybe with local modifications (`n`)).

Maybe the logic should be flipped on the if / else you added, to be more consistent with how it was (`y` = merge), but not a big deal for me.

**Comment by TomAugspurger** [comment]
Need to remember to update https://github.com/pandas-dev/pandas/wiki/Merging-Commits

Also should have have a section there on commit message format (which I messed up with b070d519c). Can link to http://pandas.pydata.org/pandas-docs/stable/contributing.html#committing-your-code

**Comment by jorisvandenbossche** [comment]
> Just to make sure, the goal of this is to allow you to push your local changes back up to the PR branch, correct? 

Yes, indeed, that's what I have used it for. I would prefer if we use the github green button more (but we can talk about in the meeting)

For that reason I put the update as 'yes' :), but can also switch or make it a different command ('update-pr.py')

**Comment by jorisvandenbossche** [comment]
> Also should have have a section there on commit message format (which I messed up with b070d51). 

What is wrong with it?

**Comment by TomAugspurger** [comment]
> What is wrong with it?

Forgot the `DOC` code in the commit message title. Pretty minor :)

File: scripts/merge-pr.py [hunk 1]
```diff
@@ -257,8 +299,13 @@ def fix_version_from_branch(branch, versions):
 print("\n=== Pull Request #%s ===" % pr_num)
 print("title\t%s\nsource\t%s\ntarget\t%s\nurl\t%s"
       % (title, pr_repo_desc, target_ref, url))
-continue_maybe("Proceed with merging pull request #%s?" % pr_num)
+continue_maybe("Proceed with updating or merging pull request #%s?" % pr_num)
 
 merged_refs = [target_ref]
 
-merge_hash = merge_pr(pr_num, target_ref)
+update = continue_maybe2("Update PR and push to github (y) or merge locally (n)?")
```
Comment (jreback) [inline scripts/merge-pr.py]: let's make this explicity, maybe ``r`` (remote) or ``l`` (local) or ``q/n`` to do nothing.

Comment (jreback) [inline scripts/merge-pr.py]: any thoughts on this?

Comment (jorisvandenbossche) [inline scripts/merge-pr.py]: that's good, will upate

**Comment by jorisvandenbossche** [comment]
@jreback What you mentioned on the hangout about the commit message: as far as I can see, you can edit both commit title and longer commit message. What else do you want to edit?

File: scripts/merge-pr.py [hunk 2]
```diff
@@ -257,8 +299,17 @@ def fix_version_from_branch(branch, versions):
...
-merge_hash = merge_pr(pr_num, target_ref)
+print("\nProceed with updating or merging pull request #%s?" % pr_num)
+update = input("Update PR and push to remote (r), merge locally (l), "
+               "or do nothing (n) ?")
+update = update.lower()
+
+if update == 'r':
+    merge_hash = update_pr(pr_num, user_login, base_ref)
+elif update == 'l':
+    merge_hash = merge_pr(pr_num, target_ref)
```
Comment (TomAugspurger) [inline scripts/merge-pr.py]: Missing an `else` block? or an `elif update == 'n'`?

Comment (TomAugspurger) [inline scripts/merge-pr.py]: Oh nevermind, that's do nothing :)

**Comment by TomAugspurger** [comment]
I just did a test, and it seems like you can edit both the title and the
squashed commit message when using the squash and merge button.

**Comment by jreback** [comment]
so the script captures the top-of-the PR, which I *usually* want some of it because its a nice sentence or 2 on what is going on this PR. The button *only* includes the title.

BUT since we have https://github.com/pandas-dev/pandas/blob/master/.github/PULL_REQUEST_TEMPLATE.md, I have to edit this out most of the time (so maybe we should just remove this entirely).

**Comment by jreback** [comment]
here's what you get if you don't explicity edit things with the scripts. Not picking on you @TomAugspurger I accidently do this too :>

```
commit ba30e3a2e376035549b009079d44ba5ca7a4c48f
Author: alexandercbooth <alexander.c.booth@gmail.com>
Date:   Wed Apr 5 16:47:07 2017 -0500

    BUG: addresses #14855 by fixing color kwarg conflict
    
     - [x] closes #14855    - [x] tests passed   - [x] passes ``git diff
    upstream/master | flake8 --diff``
    
    Author: alexandercbooth <alexander.c.booth@gmail.com>
    
    This patch had conflicts when merged, resolved by
    Committer: Tom Augspurger <tom.w.augspurger@gmail.com>
    
    Closes #14871 from alexandercbooth/fix-color-scatterm-bug and squashes the following commits:
    
    3245f09b9 [alexandercbooth] DOC: moving whatsnew entry to 0.20.0
    8ff5f51f1 [alexandercbooth] BUG: addresses #14855 by fixing color kwarg conflict

```

**Comment by jorisvandenbossche** [comment]
I understand that copying from the first post into the commit message box is some extra work, but if you typically still need to edit the generated commit message by the script, it does not seem more work?

**Comment by jreback** [comment]
> I understand that copying from the first post into the commit message box is some extra work, but if you typically still need to edit the generated commit message by the script, it does not seem more work?

what are you suggesting?

I think editing the commit messages if using the script is fine. just noting it.

I use the script because I often need to edit very minor things (like the formatting or something) in the whatsnew. Its just easier to do this right before merge.

**Comment by jorisvandenbossche** [comment]
> what are you suggesting?

What I said on the meeting: using the merge button. You can still use the script to do minor things (that's the reason I the update part), but instead of merging locally and pushing, pushing to github and directly merging there (which means editing the commit message on github instead of locally the generated one -> and that is what I meant with not much difference in work as both typically needs editing)

**Comment by jreback** [comment]
@jorisvandenbossche ahh yes I see what you mean now.

ok, then I agree the default should actually be to push back to the user PR with any changes. (default is to do nothing I think).

**Comment by jreback** [comment]
thanks @jorisvandenbossche let me give this a try!

**Pull request merged**